bf16-bitfury16.c

```diff
@@ -0,0 +1,1190 @@
+#include <arpa/inet.h>
+
+#include "miner.h"
+
+#include "bf16-bitfury16.h"
+#include "bf16-communication.h"
+#include "bf16-ctrldevice.h"
+
+#include "math.h"
+#include "sha2.h"
+#include "driver-bitfury16.h"
+
+//#define FLIP_BITS
+
+bf_cmd_description_t cmd_description[CHIP_CMD_NUM] = {
+	{
+		.cmd_code        = CHIP_CMD_TASK_STATUS,
+		.cmd_description = "CHIP_CMD_TASK_STATUS"
+	},
+	{
+		.cmd_code        = CHIP_CMD_TASK_WRITE,
+		.cmd_description = "CHIP_CMD_TASK_WRITE"
+	},
+	{
+		.cmd_code        = CHIP_CMD_TASK_SWITCH,
+		.cmd_description = "CHIP_CMD_TASK_SWITCH"
+	},
+	{
+		.cmd_code        = CHIP_CMD_READ_NONCE,
+		.cmd_description = "CHIP_CMD_READ_NONCE"
+	},
+	{
+		.cmd_code        = CHIP_CMD_SET_CLOCK,
+		.cmd_description = "CHIP_CMD_SET_CLOCK"
+	},
+	{
+		.cmd_code        = CHIP_CMD_TOGGLE,
+		.cmd_description = "CHIP_CMD_TOGGLE"
+	},
+	{
+		.cmd_code        = CHIP_CMD_SET_MASK,
+		.cmd_description = "CHIP_CMD_SET_MASK"
+	},
+	{
+		.cmd_code        = CHIP_CMD_CREATE_CHANNEL,
+		.cmd_description = "CHIP_CMD_CREATE_CHANNEL"
+	}
+};
+
+char* get_cmd_description(bf_cmd_code_t cmd_code) {
+	uint8_t i;
+	for (i = 0; i < CHIP_CMD_NUM; i++)
+		if (cmd_description[i].cmd_code == cmd_code)
+			return cmd_description[i].cmd_description;
+
+	return NULL;
+}
+
+static void shift_bits(uint8_t* data, uint8_t size, uint8_t nbits) {
+	uint8_t i;
+	uint8_t bytes = nbits / 8;
+	uint8_t bits  = nbits % 8;
+
+	for (i = 0; i < size; i++) {
+		data[i] = (data[i + bytes] << bits);
+		uint8_t minor = (data[i + bytes + 1] >> (8 - bits));
+		data[i] |= minor;
+	}
+}
+
+static uint8_t extra_bytes(uint8_t depth) {
+	return (depth * 3) / 8 + 1;
+}
+
+static uint8_t analyze_rx_data(bf_command_t* command, bf_cmd_status_t* cmd_status, uint32_t* nonces) {
+	uint8_t i;
+	uint8_t res = 0;
+
+	if (command->cmd_code & CHIP_CMD_READ_NONCE) {
+		shift_bits(command->rx, 49 + 2 + extra_bytes(command->depth), command->depth * 3);
+
+		command->status	   = command->rx[0];
+		cmd_status->status = command->rx[0];
+		command->checksum             = command->rx[1];
+		cmd_status->checksum_expected = command->checksum;
+
+		if (nonces == NULL)
+			return 1;
+
+		/* fill nonces buffer */
+		uint32_t nonce = 0x00000000;
+		for (i = 0; i <= 48; i++) {
+			if ((i % 4 == 0) && (i != 0)) {
+				if ((nonce & 0x0fffffff) != 0x0fffffff)
+					nonce ^= 0xaaaaaaaa;
+				nonces[i/4 - 1] = nonce;
+
+				if (i == 48)
+					break;
+
+				nonce = 0x00000000;
+			}
+
+			nonce |= (command->rx[i + 2] << 8*(4 - (i%4) - 1));
+		}
+
+#if 0
```

Comment:
I have left this code untouched as driver is still in development - and communication debug is sometimes needed. If you insist on removing this code - I will do this